central/clusterinit/store/store.go

```diff
@@ -145,3 +166,108 @@ func (w *storeImpl) Revoke(ctx context.Context, id string) error {
 	}
 	return nil
 }
+
+// InitiateClusterRegistration checks if another registration using the CRS with the provided CRS ID is possible.
+// If the provided id belongs to an init bundle, then registrations is always allowed, without any bookkeeping.
+func (w *storeImpl) InitiateClusterRegistration(ctx context.Context, initArtifactId, clusterName string) error {
+	w.uniqueUpdateMutex.Lock()
+	defer w.uniqueUpdateMutex.Unlock()
+
+	log.Infof("Attempting registration for cluster %s using init artifact %s", clusterName, initArtifactId)
+	initArtifactMeta, err := w.Get(ctx, initArtifactId)
+	if err != nil {
+		return errors.Wrapf(err, "retrieving init artifact meta data for ID %q", initArtifactId)
+	}
+
+	if initArtifactMeta.GetIsRevoked() {
+		log.Warnf("Init artifact %s is revoked, registration of cluster %s not allowed", initArtifactId, clusterName)
+	}
+
+	switch initArtifactMeta.Version {
+	case storage.InitBundleMeta_CRS:
+	case storage.InitBundleMeta_INIT_BUNDLE:
+		return nil
+	}
+
+	crsMeta := initArtifactMeta
+	maxRegistrations := crsMeta.MaxRegistrations
+	registrationsInitiatedSet := set.NewStringSet(crsMeta.RegistrationsInitiated...)
+	registrationsCompletedSet := set.NewStringSet(crsMeta.RegistrationsCompleted...)
+	numRegistrationsTotal := uint64(len(registrationsInitiatedSet) + len(registrationsCompletedSet))
+	if maxRegistrations > 0 && numRegistrationsTotal >= maxRegistrations {
+		log.Warnf("maximum number of cluster registrations (%d/%d) with the provided cluster registration secret %s/%q reached",
+			numRegistrationsTotal, maxRegistrations,
+			crsMeta.Id, crsMeta.Name)
+		return errors.New("maximum number of allowed cluster registrations reached")
+	}
+	if registrationsCompletedSet.Contains(clusterName) {
+		return errors.Errorf("cluster %s already registered with cluster registration secret %s/%q", clusterName, crsMeta.Id, crsMeta.Name)
+	}
+	if !registrationsInitiatedSet.Contains(clusterName) {
+		_ = registrationsInitiatedSet.Add(clusterName)
+		crsMeta.RegistrationsInitiated = registrationsInitiatedSet.AsSlice()
+		if err := w.store.Upsert(ctx, crsMeta); err != nil {
+			return errors.Wrapf(err, "updating meta data for cluster registration secret %s/%q", crsMeta.Id, crsMeta.Name)
+		}
+	}
```

Comment:
It might be a good idea to warn about repeated attempts to register an already-initiated registration, just for future troubleshooting.